nixos/syncthing: turn into RFC42 style settings

Lassulus opened the pull request:
###### Description of changes

alternative to https://github.com/NixOS/nixpkgs/pull/233377

fix https://github.com/NixOS/nixpkgs/issues/232679

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Lassulus committed:
nixos/syncthing: fix syncthing-init running by default

also remove empty values from the config

ncfavier reviewed (changes requested):
This looks ad-hoc, fragile, and inefficient. I don't like it.

doronbehar reviewed (approved):
I think this is the best possible fix proposed so far.

doronbehar commented:
> Though this fix is going to break the people who expected/use those defaults, isn't it?

I don't think so. Syncthing retains settings between reboots, and those users will just continue have the same settings, but they will be defined imperatively from now on.

ncfavier commented:
How does syncthing handle `null` values?

Lassulus commented:
> How does syncthing handle `null` values?

it should ignore them because they are invalid. Since I couldn't find anything in the docs I just tested this manually:

I first took the default config and looked at the value options.limitBandwidthInLan: which is false.

I dumped the config, changed the value to null, PUT the new config, was still false. then I changed it to true, PUT it, was true afterwards, after PUTing null again it was still true. So this seems like the perfect behavior for us

ncfavier commented:
We'll revisit this when there's no emergency. Reverted for now.

Lassulus committed:
Revert "Merge pull request #233377 from ncfavier/revert-226088"

This reverts commit 7b28ea6783baaeaebde2b58b13f141b4a6f8e050, reversing
changes made to 3009b12817c15e439609aaa794815de40a27564b.

Lassulus commented:
alright, have added a revert of the revert here so we can continue after branch off

doronbehar commented:
> > How does syncthing handle `null` values?
> 
> it should ignore them because they are invalid.

That sounds great. To further improve this, we can filter those null values prior to generating the script.

> We'll revisit this when there's no emergency. Reverted for now.

I was, and still optimistic we can reach an agreement until the branch off - we are all responsive and engaged. Hence my response is a polite :-1:.

Also, there was a lot more to be said. For instance, I think that [my PR](https://github.com/NixOS/nixpkgs/pull/230196) might have also fixed the issue for @RaitoBezarius - as it makes the script iterate each attribute in `settings` individually.

doronbehar commented on nixos/modules/services/networking/syncthing.nix:
```diff
@@ -615,7 +615,10 @@ in {
-      syncthing-init = mkIf (cfg.settings != {}) {
+      syncthing-init = mkIf (
+        # we filter all the null values out to easier compare the two sets
+        (lib.filterAttrsRecursive (_: v: v != null) cfg.settings) != { devices = {}; folders = {}; options = {}; }
```
This might fail, if in the future we'll introduce an inner option inside `settings.options` like the following:

```nix
service.syncthing.settings.options = {
    innerSyncthingOption = {
        # Default options:
        key1 = null;
        key2 = null;
    };
};
```

Ideally, I think we should create an intermediate variable that will hold all the settings, with `{}` values also completely filtered, and also inner attributes such that e.g the following settings will be transformed into simply `null`:

```nix
settings = {
    b = {
        c1 = {
            d = null;
        };
        c2 = {
            d1 = null;
            d2 = null;
        };
    };
};
```

This way, when generating the `syncthing-init.service` script, if we'll read a completely filtered version of `cfg.settings`, we will avoid relying on the API not erroring when JSON values are `null`.

Lassulus replied:
maybe we should just filter some more? how about:

```
filterAttrsRecursive (_: v: v != null && v != {} ) cfg.settings
```

Lassulus replied:
I pushed a small change to filter settings recursively (also for the json payload)

RaitoBezarius commented:
> Though I am rather sure the solution would have worked for you. 

I am certain too, but I am an advanced NixOS user, I don't care about me :P. I have to care about the others. :)

doronbehar commented:
I think this PR can be marked as ready for review, since the [`release-23.05`](https://github.com/NixOS/nixpkgs/tree/release-23.05) branch off has already been done.

RaitoBezarius commented:
Correct, please write NixOS tests to figure out the right semantics for the fix if you cannot get someone to test your changes :)

doronbehar committed:
nixosTests.syncthing-no-settings: init

doronbehar commented:
In addition to making sure in a `nix-repl` that `lib.converge` function works as expected, with `raito.nix` from [my comment at #232679](https://github.com/NixOS/nixpkgs/issues/232679#issuecomment-1556761770), I added a nixos test that makes sure that no `syncthing-init.service` service is created when no `settings` are defined. I also checked manually that if I add a `settings.options.relaysEnabled = true;` then the test I added fails.

Lassulus commented:
from my side this look ready.

ncfavier commented:
It's not. I don't have time to do a proper review now. Do not merge.

ncfavier commented:
If you want this to go anywhere, please bring this PR down to the bare minimum needed to support arbitrary settings. A couple `freeformType`s should do, we can think about reorganising the module and when exactly `syncthing-init` should be enabled later.

`converge (filterAttrsRecursive ...` is an abomination and I'm a bit worried that everyone seems fine with it.

Lassulus commented:
> 

thats generally used quite often with freeformTypes, like in home-assistant, keycloak, bookstack. I don't see any upside moving folders and devices out of settings again, since the extraConfig gives the same functionality anyway and causes weird double configuration scenarios. I could rework it anyway but it seems like a bit of redundant work.

doronbehar commented:
> thats generally used quite often with freeformTypes, like in home-assistant, keycloak, bookstack.

By "thats" you mean `converge (filterAttrsRecursive`?

Lassulus commented:
> > thats generally used quite often with freeformTypes, like in home-assistant, keycloak, bookstack.
> 
> By "thats" you mean `converge (filterAttrsRecursive`?

ah yeah :D , for some reason my quote reply failed.

Lassulus commented:
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/home-automation/home-assistant.nix#L14

https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/video/frigate.nix#L21

https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/monica.nix#L408

https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/snipe-it.nix#L429

https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/keycloak.nix#L471

https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/security/kanidm.nix#L6

https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/bookstack.nix#L394

https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/monitoring/parsedmarc.nix#L472

are the ones I could find

Lassulus commented:
@ncfavier do you have a rough estimate when you will come around to review this or https://github.com/NixOS/nixpkgs/pull/234262 ? right now this is in limbo and I'm not aware of any more regressions with this PR

ncfavier commented:
@infinisil may I ask for your opinion? I'm not sure how to best implement RFC 42 here; the proposed solutions leave me deeply unsatisfied and I can't seem to articulate why or suggest alternatives.

Particular points I'm iffy about:
- moving `devices`, `folders` option trees down under `settings` (they seem like important options that should be easily accessible)
- removing `watch`, `watchDelay`, `rescanInterval` in favour of their API-internal names `fsWatcherEnabled`, `fsWatcherDelayS`, `rescanIntervalS` (ditto, I guess)
- the horrible `cleanedConfig` hack; I've suggested looking at `options.services.syncthing.settings.isDefined` before to determine whether to enable the syncthing-init service, does that seem like a better idea?

Lassulus commented:
Just to write down my thought process and different tradeoffs I thought about while implementing these changes:
> 
>     * moving `devices`, `folders` option trees down under `settings` (they seem like important options that should be easily accessible)

since `.settings` is a freeFormType and describes the API endpoint, having these outside of settings would cause some weird behavior where people could define folders/devices twice, once through `services.syncthing.folders` and through `services.syncthing.settings.folders`.
We could strip the folders from settings but that would make it more obscure and we would stray away from the API.

Ideally I want people to dump their current config and just put that into settings to make it declarative.

 
>     * removing `watch`, `watchDelay`, `rescanInterval` in favour of their API-internal names `fsWatcherEnabled`, `fsWatcherDelayS`, `rescanIntervalS` (ditto, I guess)

this is similar to the previous issue, but I don't see a quick solution. If we turn folders into a freeForm people could define `rescanIntervalS` directly and we would need to handle to conflict with our current/previous `rescalInterval`.
But not sure which one should win in that scenario?

Also the current settings has a default which ideally would need to keep in sync with upstream if the default changes there.
Also there is no way to just ignore the value completely and have it set manually in the webinterface, since it would be override with the default value of the option. The new code would let people have folders declaratively but be able to change folder settings manually.
 
>     * the horrible `cleanedConfig` hack; I've suggested looking at `options.services.syncthing.settings.isDefined` before to determine whether to enable the syncthing-init service, does that seem like a better idea?

I looked into isDefined, but if a freeForm has options with default values (it breaks without those default values in my tests) it seems to always be true. Also stripping empty sets and nulls gives us the ability to have documentation for certain options without forcing them to a default value by simply setting them to null. Other alternatives I tried out made the check more complex and gave us less features with more edgecases to handle in the future. Like when people add more option.

infinisil commented:
[Part 2 of RFC 42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md#part-2-1) encourages adding extra options beside `settings` when it makes sense. So I agree with @ncfavier that `devices`, `watch`, etc. would make sense to keep as options.

@Lassulus it shouldn't be a problem if these extra options are implemented in terms of `settings`, something like this:

```
settings.fsWatcherDelayS = cfg.watchDelay;
```

> the horrible `cleanedConfig` hack; I've suggested looking at `options.services.syncthing.settings.isDefined` before to determine whether to enable the syncthing-init service, does that seem like a better idea?

I think `isDefined` is something that should never be used for NixOS modules, because it leads to cases where things can change depending on the priority of options, even though all options are set to the same value, which is very unintuitive.

Most likely this is also the reason upstream why the `cleanedConfig` hack is necessary here, because upstream config behavior changes depending on whether the default is used or the value is set explicitly, even though the values might be the same. This could be caused by the upstream config not providing any way to set the default.

So I think ideally this should be fixed in upstream so that we don't need the `cleanedConfig` hack here, but until then I think it's fine to have it. I like it better than having to rely on `isDefined`.

ncfavier commented:
(`isDefined` wouldn't work anyway, as @Lassulus points out, so we would have to use something like `options.services.syncthing.settings.highestPrio > (mkOptionDefault {}).priority`, but I imagine your concerns apply all the same?)

doronbehar commented:
BTW I'd really like to get this merged so that I can get https://github.com/NixOS/nixpkgs/pull/230196 going as well...

doronbehar commented:
> I didn't understand the point of configuring settings in terms of their aliases

Me too, but if others will insist, perhaps we can force the user to use (in this example) `watchDelay` and not `fsWatcherDelayS`, by triggering a warning or `throw`ing an error if the later is used.

I'll be able to live with the above suggestion, but I still think it's not too bad to ask our users to change some of these settings in order to conform to upstream's terminology - making us and the NixOS modules' users' maintenance burden easier

doronbehar commented:
Rebased, and solved merge conflicts after https://github.com/NixOS/nixpkgs/pull/239427 , hopefully without mistakes.

doronbehar commented:
@ofborg test syncthing-no-settings

doronbehar commented:
So what is the status of this?

Lassulus commented:
well I tried to bump some people but basically we are in review hell since @nbp still seems unhappy with the solution? not sure exactly how to proceed. But the current solution still seems like the best we can come up with. having devices and folders outside of settings would be misleading since they end up in settings anyway. and discoverability should be the same since documentation is generated either way?

I would be very happy if we could merge this and fix stuff later if it comes up. but even the last regressions before the channel bump were mainly because syncthing-init ran in weird circumstances where it should not.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/62

ncfavier commented:
How about making the type of `settings` `nullOr (submodule ...)` and defaulting to `null`? That way the check is just `cfg.settings == null`.

Lassulus commented:
Ah we tried that already but the default value of settings gets overriden by the default values of the sub options

Lassulus commented:
since communication with @ncfavier seems kinda stuck (please don't take it as an offence) and there are no other voices against merging this, I will self merge this. We are already like 1/3 of the time to a new release and there seems no forward movement and the closer we come to a release the bigger the fallout we have to fix last minute if something unexpected happens (again?)
